File: Data Structure/Queue/go/queue.go
package main

import "fmt"

type Queue struct {
	queue []int
}

func NewQueue() *Queue {
	return &Queue{}
}

func (q *Queue) isEmpty() bool {
	return q.size() == 0
}

func (q *Queue) size() int {
	return len(q.queue)
}

func (q *Queue) push(val int) {
	q.queue = append(q.queue, val)
}

func (q *Queue) pop() int {
	if q.isEmpty() {
		panic("Queue is Empty")
	}
	size := q.size()
	element := q.queue[0]
	if size != 0 {
		q.queue = q.queue[1:size]
	} else {
		q.queue = q.queue[:size]
	}

	return element
}

func (q *Queue) print() {
	if q.isEmpty() {
		panic("Queue is Empty")
	}
	for _, v := range q.queue {
		fmt.Println(v)
	}
}

func main() {
	queue := NewQueue()
	fmt.Println(queue.isEmpty())
	queue.push(1)
	queue.push(2)
	queue.push(3)
	fmt.Println("size: ", queue.size())
	fmt.Println(queue.isEmpty())
	queue.print()
	fmt.Println("=============Pop================")
	fmt.Println(queue.pop())
	fmt.Println(queue.pop())
	fmt.Println("=============Push================")
	queue.push(4)
	queue.push(5)
	fmt.Println("size: ", queue.size())
	queue.print()
}

File: Data Structure/Queue/java/Queue.java
package queue;

public class Queue {
    public static void main(String[] args) {
        CircleArrayQueue c = new CircleArrayQueue(5);
        c.add(10);
        c.displayQueue();
        c.add(20);
        c.displayQueue();
        System.out.println(c.peek());
        System.out.println(c.peek());
        System.out.println(c.element());
        System.out.println(c.peek());
        c.displayQueue();
    }
}

class ArrayQueue {
    private int maxSize;
    private int front; // Queue header
    private int rear; // Queue tail
    private int[] arr;

    public ArrayQueue(int maxSize) {
        this.maxSize = maxSize;
        arr = new int[maxSize];
        front = -1;
        rear = -1;
    }

    public boolean isFull() {
        return rear == (maxSize - 1);
    }

    public boolean isEmpty() {
        return rear == front;
    }

    public void add(int n) {
        if(isFull()) {
            throw new RuntimeException("is full");
        }
        rear++;
        this.arr[rear] = n;
    }

    public int element() {
        if(isEmpty()) {
            throw new RuntimeException("is empty");
        }
        front++;
        return arr[front];
    }

    public int peek(){
        if(isEmpty()) {
            throw new RuntimeException("is empty");
        }
        return arr[front+1];

    }

    public void display() {
        if(isEmpty()) {
            throw new RuntimeException("is empty");
        }
        for (int i=0; i<arr.length; i++){
            System.out.printf("arr[%d]=%d\n", i, arr[i]);
        }
    }

    public int size() {
        if(isEmpty()) return 0;
        return rear+1;
    }
}

class CircleArrayQueue {
    private int maxSize;
    private int front; // Queue header
    private int rear; // Queue tail
    private int[] arr;

    public CircleArrayQueue(int maxSize) {
        this.maxSize = maxSize;
        arr = new int[maxSize];

    }
    
    public boolean isFull() {
        return (rear+1)%maxSize == front;
    }

    public boolean isEmpty() {
        return rear == front;
    }

    public void add(int n) {
        if(isFull()) {
            throw new RuntimeException("is full");
        }
        arr[rear] = n;
        rear = (rear+1)%maxSize;
    }

    public int element(){
        if(isEmpty()) {
            throw new RuntimeException("is empty");
        }
        int val = arr[front];
        front = (front+1)%maxSize;
        return val;
    }


    public void displayQueue() {
        if(isEmpty()) {
            throw new RuntimeException("is empty");
        }
        for(int i=front; i<(front+ size()); i++) {
            System.out.printf("arr[%d]=%d\n", i%maxSize, arr[i%maxSize]);
        }
    }

    public int size() {
        return (rear+maxSize-front)%maxSize;
    }

    public int peek() {
        if(isEmpty()) {
            throw new RuntimeException("is empty");
        }
        return arr[front];
    }
}